### PR TITLE
HHH-16316 Move version specific dialects of unsupported versions to hibernate-community-dialects

### PR DESCRIPTION
**Pull request by beikov**
https://hibernate.atlassian.net/browse/HHH-16316

### CODE REVIEW

File: hibernate-core/src/main/java/org/hibernate/boot/registry/selector/internal/DefaultDialectSelector.java [hunk 1]
```diff
@@ -112,75 +89,85 @@ public Class<? extends Dialect> resolve(final String name) {
...
 			case "SybaseASE15":
-				return SybaseASE15Dialect.class;
 			case "SybaseASE157":
-				return SybaseASE157Dialect.class;
+				return findCommunityDialect( name );
 		}
 		return null;
 	}
 
+	private static Class<? extends Dialect> findCommunityDialect(String name) {
```
Comment (sebersole) [inline hibernate-core/src/main/java/org/hibernate/boot/registry/selector/internal/DefaultDialectSelector.java:154]: Still not entirely sure about exception v. warning here.  As I understand, the incoming `name` is a value explicitly set for which Dialect to use.

What's the ultimate outcome in this case?  We fall back to database-metadata resolution I assume?

File: documentation/src/main/asciidoc/userguide/chapters/domain/inheritance.adoc [hunk 2]
```diff
@@ -313,24 +313,6 @@ include::{extrasdir}/entity-inheritance-table-per-class-query-example.sql[]
...
-- `H2Dialect`
-- `HSQLDialect`
-- `Ingres9Dialect`
-- `MySQL5Dialect`
-- `Oracle8iDialect`
-- `Oracle9Dialect`
-- `PostgreSQL81Dialect`
-- `RDMSOS2200Dialect`
-====
-
```
Comment (yrodiere) [inline documentation/src/main/asciidoc/userguide/chapters/domain/inheritance.adoc:333]: Was that on purpose? It seems unrelated and breaks formatting (you're not closing the admonition block).

File: hibernate-core/src/main/java/org/hibernate/boot/registry/selector/internal/DefaultDialectSelector.java (hunk 1 above)
Comment (yrodiere) [reply hibernate-core/src/main/java/org/hibernate/boot/registry/selector/internal/DefaultDialectSelector.java:154]: +1 for exception. If something gets explicitly requested and we can't find it, there's a problem that should be addressed right now.

Falling back to database-metadata resolution is unlikely to help if the user selected a community (i.e. unsupported) dialect in the first place.

File: hibernate-core/src/main/java/org/hibernate/engine/jdbc/dialect/internal/DialectFactoryImpl.java [hunk 3]
```diff
@@ -124,6 +153,20 @@ private Dialect constructDialect(Object dialectReference, DialectResolutionInfoS
...
+			final String dialectFqn = dialectReference.toString();
+			if ( LEGACY_DIALECTS.contains( dialectFqn ) ) {
+				throw new StrategySelectionException(
+						"Couldn't load the dialect class for the `hibernate.dialect` [" + dialectFqn + "], " +
+								"because the application is missing a dependency on the hibernate-community-dialects module. " +
+								"Hibernate 6.2 dropped support for database versions that are unsupported by vendors  " +
+								"and code for old versions was moved to the hibernate-community-dialects module. " +
+								"For further information, read https://in.relation.to/2023/02/15/hibernate-orm-62-db-version-support/",
+						e
+				);
```
Comment (yrodiere) [inline hibernate-core/src/main/java/org/hibernate/engine/jdbc/dialect/internal/DialectFactoryImpl.java:166]: :+1: great to give that info to users.

File: hibernate-core/src/test/java/org/hibernate/orm/test/strategyselectors/DefaultDialectSelectorTest.java [hunk 4]
```diff
@@ -21,8 +25,6 @@ public class DefaultDialectSelectorTest {
 	@Test
 	public void verifyAllDialectNamingResolve() {
 		testDialectNamingResolution( DB2Dialect.class );
-		testDialectNamingResolution( DB2390Dialect.class );
-		testDialectNamingResolution( DB2390V8Dialect.class );
```
Comment (yrodiere) [inline hibernate-core/src/test/java/org/hibernate/orm/test/strategyselectors/DefaultDialectSelectorTest.java:25]: Should there be a dedicated test in the `hibernate-community-dialects` module, then? I understand we don't guarantee that dialects work, but I'd expect us to guarantee that name resolution, at least, does?

**Review by yrodiere — COMMENTED** [review]
Thanks, a few comments below.

Please don't forget to update migration guides ;D

File: documentation/src/main/asciidoc/userguide/chapters/domain/inheritance.adoc (hunk 2 above)
Comment (beikov) [reply documentation/src/main/asciidoc/userguide/chapters/domain/inheritance.adoc:333]: I'll fix the block closing, but the section is unnecessary, because all supported dialect versions support `union all` now.

File: hibernate-core/src/main/java/org/hibernate/boot/registry/selector/internal/DefaultDialectSelector.java (hunk 1 above)
Comment (beikov) [reply hibernate-core/src/main/java/org/hibernate/boot/registry/selector/internal/DefaultDialectSelector.java:154]: Ok, then I'll change to throwing an exception.

File: hibernate-core/src/test/java/org/hibernate/orm/test/strategyselectors/DefaultDialectSelectorTest.java (hunk 4 above)
Comment (beikov) [reply hibernate-core/src/test/java/org/hibernate/orm/test/strategyselectors/DefaultDialectSelectorTest.java:25]: I'll move the test assertions into that module.
